0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,179 @@
+
+## Motivation
+
+The initial [Ethereum Crypto Primitive PRD](https://www.notion.so/minaprotocol/Ethereum-Primitives-Support-PRD-d89af720e1c94f7b90166709432e7bd5) describes the importance of establishing cryptographic compatibility with Ethereum. The [ECDSA PRD](https://www.notion.so/minaprotocol/ECDSA-ver-gadget-PoC-PRD-9458c38adf204d6b922deb8eed1ac193) and the
+[Keccak PRD](https://www.notion.so/minaprotocol/Keccak-gadget-PoC-PRD-59b024bce9d5441c8a00a0fcc9b356ae) then go into detail and describe two of the most important building blocks to achieve Ethereum compatibility in a cryptographic sense.
+
+This RFC describes the steps needed to expose both ECDSA and SHA3/Keccak to SnarkyJS, enabling SnarkyJS developer to use these primitives in their applications and smart contracts.
+
+It is important to mention that this RFC only considers the work required to _expose_ already existing cryptographic primitives (gates and gadgets), which have previously been implemented by the crypto team, from the OCaml bindings layer to SnarkyJS - it is not required to implement additional cryptographic primitives. Changes will only impact [snarkyjs-bindings](https://github.com/o1-labs/snarkyjs-bindings) and [SnarkyJS](https://github.com/o1-labs/snarkyjs) itself.
+
+Once completed, SnarkyJS users will be able to leverage ECDSA and SHA3/Keccak to build applications that integrate with Ethereum and other use cases that require the use of said cryptographic primitives.
+
+## Detailed design
+
+### SHA3/Keccak
+
+The Keccak and SHA3 gadget has been implemented by the crypto team, enabling us to leverage the already existing bindings layer to and from OCaml. This design allows us to simply integrate the new gadgets into SnarkyJS by simply exposing them in `ocaml/lib/snarky_js_bindings_lib.ml`.
+
+For Keccak/SHA3, the implementation exposes two ready-to-go functions.
+
+```ocaml
+(** Gagdet for NIST SHA-3 function for output lengths 224/256/384/512 *)
+val nist_sha3 :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> int
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+(** Gadget for Keccak hash function for the parameters used in Ethereum *)
+val eth_keccak :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+```
+
+These two functions will be imported into the bindings layer and exposed via a new sub-module as part of the already existing `Snarky` module.
+
+```ocaml
+
+  module Sha = struct
+    let create message nist length =
+      let message_array = Array.to_list message in
+      if Js.to_bool nist then
```

Comment:
Is `nist` meant to be the flag in the Snarky gadget? If so, after being asked in review of the Keccak PR, it got renamed into `nist_version`, still a boolean.